replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -13,140 +30,103 @@ private[journal] sealed abstract class Batch extends Product {
+  private object State {
+    def apply(records: NonEmptyList[ActionRecord[Action]]): State = {
+      records.reverse.foldLeft(State()) { _.handle(_) }
     }
+  }
 
-    def dropDeleted(appends: Appends, delete: Action.Delete): Option[Appends] = {
-      // we cannot drop individual events in record as we do not parse the payload here (`records.head.action.payload: ByteVector`)
-      val validRecords = appends.records.collect { case record if record.action.range.to >= delete.to.value => record }
-      NonEmptyList.fromList(validRecords).map { Appends(appends.offset, _) }
-    }
+  private final case class State(
+    private val purge: Option[Purge]     = None,
+    private val appends: Option[Appends] = None,
+    private val delete: Option[Delete]   = None,
+  ) {
+    // Expects records to be provided in reversed order, e.g., youngest first
+    private def handle: ActionRecord[Action] => State = {
+      case _ if this.purge.nonEmpty => // ignore all actions before `Purge`
+        this
 
-    records
-      .foldLeft(List.empty[Batch]) { (bs, record) =>
-        val offset = record.partitionOffset.offset
+      case ActionRecord(_: Action.Mark, _) =>
+        this
 
-        def appendsOf(records: NonEmptyList[ActionRecord[Action.Append]]): Appends = {
-          Appends(offset, records)
-        }
+      case ActionRecord(purge: Action.Purge, partitionOffset: PartitionOffset) =>
+        handlePurge(purge, partitionOffset)
 
-        def deleteOf(to: DeleteTo, origin: Option[Origin], version: Option[Version]): Delete = {
-          Delete(offset, to, origin, version)
-        }
+      case ActionRecord(delete: Action.Delete, partitionOffset: PartitionOffset) =>
+        handleDelete(delete, partitionOffset)
 
-        def purgeOf(origin: Option[Origin], version: Option[Version]): Purge = {
-          Purge(offset, origin, version)
-        }
+      case ActionRecord(append: Action.Append, partitionOffset: PartitionOffset) =>
+        handleAppend(append, partitionOffset)
+    }
 
-        def actionRecord[A <: Action](a: A): ActionRecord[A] = {
-          record.copy(action = a)
-        }
+    private def handlePurge(purge: Action.Purge, partitionOffset: PartitionOffset): State = {
+      this.copy(
+        purge = Purge(partitionOffset.offset, purge.origin, purge.version).some,
+      )
+    }
 
-        def origin: Option[Origin] = {
-          bs.foldRight(none[Origin]) { (b, origin) =>
-            origin orElse {
-              b match {
-                case b: Batch.Delete  => b.origin
-                case _: Batch.Appends => none
-                case b: Batch.Purge   => b.origin
-              }
-            }
-          }
-        }
+    private def handleDelete(delete: Action.Delete, partitionOffset: PartitionOffset): State = {
+      val delete_ = this.delete match {
+        case Some(younger) =>
+          // take `origin` and `version` from "older" entity, if it has them
```

Comment:
Please elaborate _why_ `origin` and `version` are taken from "older" entity

Reply:
from what I understood - we want to save the `origin` and `version` of *first* action, which creates the new journal (strange, that original test cases do not check if `append`s keep it correctly - maybe in initial version `delete` didn't have `origin` and `version`?)

Reply:
probably, the `Option` part is for backwards compatibility with older client code - eventually we should remove the optionality